WordPress/openverse
# Show timeout errors on the frontend

## Fixes



Fixes #632 by @sarayourfriend

## Description


This PR updates the `FetchState` to always use the `FetchingError` instead of `NuxtError`|`string`|`FetchingError` for the error object. 
`FetchingError` interface provides more information about the error, including the response status code, the string code of the error from axios, the kind of request (single result, search, provider or the related media), and additional details such as the item id.

This interface makes it easier to show the timeout error pages on `SERVER_TIMEOUT` or `ECONNABORTED` errors.

`FetchState` was made a generic to allow the use of `string` and `NuxtError` for the `fetchingError` value. This PR removes this to make fetchState simpler.

## Testing Instructions

To simulate the server timeout, lower the default timeout to 30:
https://github.com/WordPress/openverse/blob/026a328a781ebebfc4b29a22b4b041b05acdb544/frontend/src/data/api-service.ts#L8
Now, you should see the timeout error page any time you search for something. You should also see the same page if you refresh this search page (so, the page will be server-rendered now).

To see the no results error page, search for something like "querywithnoresults". It should work for all search types, both on CSR and SSR, and should not try to reload the page on SSR.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
